### PR TITLE
feat: max requests per socket

### LINKED ISSUE #3260: max requests on a socket
## 🚀 Feature Proposal

In keep-alive response header, there is a posibility to define the timeout and max requests for the connection
https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Keep-Alive
 
I would like propose the functionality on the server, that will count requests per socket and send "connection: closed" header and actually close the connection when it reached the maximum value

## Motivation

Currently as fast as I can understand, nothing can force the connection on the server to be closed, for example both 
"connectionTimeout" and "keepAliveTimeout" define the time of inactivity, for the socket to be closed, but if the socket is always active, or the client ignores this header, server will not send the "connection: close" header and close the connection

### PR DESCRIPTION
**Pull request by climba03003**
Fixes #3260 

~~I can see that the header is not fully implemented according to provide both `timeout` and `max` field.~~

~~The best case should be returning the header
`Keep-Alive: timeout=5, max=2`~~

#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

**Comment by mcollina** [comment]
The max part is deprecated by the spec.

File: test/maxRequestsPerSocket.test.js [hunk 1]
```diff
@@ -0,0 +1,56 @@
+'use strict'
+
+const net = require('net')
+const { test } = require('tap')
+const semver = require('semver')
+const Fastify = require('../fastify')
+
+test('maxRequestsPerSocket on node version greater than 16.10.0', { skip: semver.lt(process.versions.node, '16.10.0') }, t => {
```
Comment (climba03003) [inline test/maxRequestsPerSocket.test.js:8]: Github Actions is not updated to node version `16.10.0`.
For those who reviewing this PR should run this test in local environment.

Comment (climba03003) [reply test/maxRequestsPerSocket.test.js:8]: It is released.
https://github.com/actions/node-versions/releases/tag/16.10.0-1264673745

File: docs/Server.md [hunk 2]
```diff
@@ -82,6 +83,17 @@ is in use. Also, when `serverFactory` option is specified, this option is ignore
 
 + Default: `5000` (5 seconds)
 
+<a name="factory-max-requests-per-socket"></a>
+### `maxRequestsPerSocket`
+
+Defines the maximum number of requests socket can handle before closing keep alive connection. See documentation for
+[`server.maxRequestsPerSocket` property](https://nodejs.org/dist/latest-v16.x/docs/api/http.html#http_server_maxrequestspersocket)
+to understand the effect of this option. This option only applies when HTTP/1
```
Comment (mcollina) [inline docs/Server.md:91]:
```suggestion
to understand the effect of this option. This option only applies when HTTP/1.1
```

File: docs/Server.md [hunk 3]
```diff
@@ -82,6 +83,17 @@ is in use. Also, when `serverFactory` option is specified, this option is ignore
...
+is in use. Also, when `serverFactory` option is specified, this option is ignored.
+>  Only node version greater or equal to 16.10.0 support this option.
```
Comment (mcollina) [inline docs/Server.md:93]:
```suggestion
>  At the time of this writing, only node version greater or equal to 16.10.0 support this option. Check the Node.js documentation for availability in the version you are running.
```

File: docs/Server.md [hunk 4]
```diff
@@ -82,6 +83,17 @@ is in use. Also, when `serverFactory` option is specified, this option is ignore
...
+
++ Default: `null` (no limit)
```
Comment (mcollina) [inline docs/Server.md:95]: I would prefer to have `0` instead of `null` to indicate no limit.

Comment (climba03003) [reply docs/Server.md:95]: I need to test against the `http` server to see if `0` work as `null`.

File: lib/server.js [hunk 5]
```diff
@@ -28,6 +30,16 @@ function createServer (options, httpHandler) {
   } else {
     server = http.createServer(httpHandler)
     server.keepAliveTimeout = options.keepAliveTimeout
+    if (options.maxRequestsPerSocket > 0 && semver.lt(versions.node, '16.10.0')) {
+      // when node version smaller than 16.10.0 and is a number
+      // it is not a valid option
+      throw new FST_ERR_HTTP_MAX_REQUESTS_PER_SOCKET_INVALID_VERSION()
```
Comment (mcollina) [inline lib/server.js:36]: This will complicate things if it is backported to previous lines - ideally it can land in v14. I would recommend not having this check and just set it.

File: test/maxRequestsPerSocket.test.js [hunk 6]
```diff
@@ -0,0 +1,160 @@
+'use strict'
+
+const net = require('net')
+const { test } = require('tap')
+const semver = require('semver')
+const Fastify = require('../fastify')
+const proxyquire = require('proxyquire')
+
+test('maxRequestsPerSocket on node version >= 16.10.0', { skip: semver.lt(process.versions.node, '16.10.0') }, t => {
```
Comment (mcollina) [inline test/maxRequestsPerSocket.test.js:9]: move the skip check to the top of the file and reuse it.

File: lib/server.js (hunk 5 above)
Comment (climba03003) [reply lib/server.js]: I will keep the `options.maxRequestsPerSocket > 0` check, as setting `0` will block all incoming connection.

**Review by jsumners — APPROVED** [review]

**Review by RafaelGSS — APPROVED** [review]
LGTM. It will land as minor?

**Comment by mcollina** [comment]
Yes

File: lib/server.js [hunk 7]
```diff
@@ -28,6 +28,12 @@ function createServer (options, httpHandler) {
   } else {
     server = http.createServer(httpHandler)
     server.keepAliveTimeout = options.keepAliveTimeout
+    // we treat zero as null
+    // and null is the default setting from nodejs
+    // so we do not pass the option to server
+    if (options.maxRequestsPerSocket > 0) {
+      server.maxRequestsPerSocket = options.maxRequestsPerSocket
+    }
```
Comment (mcollina) [inline lib/server.js:36]: For all of you reading this, I just opened https://github.com/nodejs/node/issues/40190 because of it. This if should not be needed.

**Review by mcollina — APPROVED** [review]
lgtm
